### PR TITLE
Add pydocstyle linter

### PR DESCRIPTION
**Pull request by inducer**
This finds things like
```
        D417: Missing argument descriptions in the docstring (argument(s) discr are missing descriptions in 'inviscid_operator' docstring)
         D401: First line should be in imperative mood; try rephrasing (found 'The')
         D100: Missing docstring in public module
```
as well as heaps and heaps of formatting stuff.

@lukeolson @majosm @matthiasdiener @MTCam Thoughts?

### CODE REVIEW

**Comment by MTCam** [comment]
I like this, in principal.  flake8/editor integration changed my life. I have a similar hope for this tool.

Do we intend on fixing the gobs of stuff it finds?  If we can make it shut up about the stuff we don't care about and fix all the stuff we do, then it could be very useful.    On the other hand, if whatever new violations I produce with my new code are buried in a mountain of other stuff and CI passes whether they are fixed or not - then I would tend to care less about what it tells me.

If there is a mountain of busy work associated with the current report - then I propose that we work on bringing master into the quiet zone of this tool before adding it to CI.

**Comment by inducer** [comment]
> If there is a mountain of busy work associated with the current report - then I propose that we work on bringing master into the quiet zone of this tool before adding it to CI.

I agree. The main question I'm meaning to raise here is whether the signal/noise ratio of this tool looks sufficiently high to justify that effort. I'm leaning slightly yes, which is why I'm proposing it.

**Comment by MTCam** [comment]
> > If there is a mountain of busy work associated with the current report - then I propose that we work on bringing master into the quiet zone of this tool before adding it to CI.
> 
> I agree. The main question I'm meaning to raise here is whether the signal/noise ratio of this tool looks sufficiently high to justify that effort. I'm leaning slightly yes, which is why I'm proposing it.

Looks like master could be quieted pretty easily. Regardless, this seems pretty useful to me.

**Comment by inducer** [comment]
There's https://pypi.org/project/flake8-docstrings/ which integrates this with Flake8.

**Comment by MTCam** [comment]
> There's https://pypi.org/project/flake8-docstrings/ which integrates this with Flake8.

That seems mean, but I'm game :)

**Comment by inducer** [comment]
Decision: crowdsource, shepherded by @MTCam

**Comment by inducer** [comment]
FYI Resolved some merge conflicts

**Comment by MTCam** [comment]
We need to do something with this to make it accept our files.   Currently, it reports we don't have any module-level docstrings because it appears to hate __doc__ for constructing the docstring - as well as the fact that __copyright__ comes before __doc__ (i.e. assuming the __doc__ issue can be fixed).

Any ideas @inducer?

**Comment by inducer** [comment]
My take is that `__doc__` goes away and turns into an undecorated string at the start of the module. Fine by me, just a different convention.

**Comment by MTCam** [comment]
Silencing some of these complaints required my adding some pretty obvious single line documentation.

**Comment by inducer** [comment]
Thanks for working on this! LGTM, though I can't approve because I initially created this PR.

**Comment by MTCam** [comment]
Thanks for turning this on!  Adding this linter to flake8 has helped (me) with the docs.  I'm sure you noticed some of the docs weren't especially useful or informative, as we have discussed, but it seems to be worth it.

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -47,6 +47,22 @@ jobs:
...
+        -
+          uses: actions/setup-python@v1
+          with:
+            python-version: '3.x'
+        - name: Lint
+          run: |
+            python3 -m venv myenv
+            source myenv/bin/activate
+            python -m pip install wheel
+            python -m pip install pydocstyle
```
Comment (lukeolson) [inline .github/workflows/ci.yaml:63]: we could consider removing this ci section and absorb it above with `flake8-docstrings` (and possibly other flake8 extensions.

File: mirgecom/boundary.py [hunk 2]
```diff
@@ -1,3 +1,12 @@
+"""Boundary condition implementations.
+
+Boundary Conditions
+^^^^^^^^^^^^^^^^^^^
```
Comment (lukeolson) [inline mirgecom/boundary.py:4]: Why do we use an H5 header here? (there may be a good reason)

File: mirgecom/boundary.py [hunk 3]
```diff
@@ -28,28 +37,28 @@
...
     .. automethod:: __init__
     .. automethod:: boundary_pair
     """
+
     def __init__(self, userfunc):
+        """Set the boundary function.
+
+        Parameters
+        ----------
+        userfunc
```
Comment (lukeolson) [inline mirgecom/boundary.py:52]: no type, but we're not checking for this yet.  skip.

**Review by lukeolson — APPROVED** [review]

**Pull request merged**